services/venice-server/src/test/java/com/linkedin/venice/listener/StorageReadRequestHandlerTest.java

```diff
@@ -621,4 +622,28 @@ public void testMisRoutedStoreVersion() throws Exception {
     verify(context).writeAndFlush(shortcutResponseArgumentCaptor.capture());
     Assert.assertTrue(shortcutResponseArgumentCaptor.getValue().isMisroutedStoreVersion());
   }
+
+  @Test
+  public void testNoStorageEngineReturn503() throws Exception {
+    String storeName = "testStore";
+    String keyString = "key_byte";
+    GetRouterRequest request = mock(GetRouterRequest.class);
+    doReturn(false).when(request).shouldRequestBeTerminatedEarly();
+    doReturn(SINGLE_GET).when(request).getRequestType();
+    doReturn(Version.composeKafkaTopic(storeName, 1)).when(request).getResourceName();
+    doReturn(storeName).when(request).getStoreName();
+    Store store = mock(Store.class);
+    doReturn(Optional.empty()).when(store).getVersion(anyInt());
+    doReturn(store).when(storeRepository).getStore(storeName);
+    doReturn(1).when(store).getCurrentVersion();
+    when(storageEngine.isClosed()).thenReturn(true);
+    doReturn(null).when(storageEngineRepository).getLocalStorageEngine(any());
+    doReturn(keyString.getBytes()).when(request).getKeyBytes();
+    StorageReadRequestHandler requestHandler = createStorageReadRequestHandler();
+    requestHandler.channelRead(context, request);
+    ArgumentCaptor<HttpShortcutResponse> shortcutResponseArgumentCaptor =
+        ArgumentCaptor.forClass(HttpShortcutResponse.class);
+    verify(context).writeAndFlush(shortcutResponseArgumentCaptor.capture());
+    Assert.assertEquals(shortcutResponseArgumentCaptor.getValue().getStatus(), SERVICE_UNAVAILABLE);
```

Comment:
Can we validate the `BAD_REQUEST` code path as well by this unit test?

Reply:
Sure